File: java/src/main/java/io/dure/tree/RecoverTree.java
package io.dure.tree;

import java.util.Deque;
import java.util.LinkedList;

public class RecoverTree {
    public void recoverTree(TreeNode root) {
        Deque<TreeNode> stack = new LinkedList<>();
        TreeNode curr = root;
        TreeNode prev = null, x = null, y = null;
        while (curr != null || !stack.isEmpty()) {
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }
            curr = stack.pop();
            if (prev != null && prev.val > curr.val) {
                y = root;
                if (x == null) {
                    x = curr;
                } else {
                    break;
                }
            }
            prev = curr;
            curr = curr.right;
        }
    }

    private void swap(TreeNode x, TreeNode y) {
        int temp = x.val;
        x.val = y.val;
        y.val = temp;
    }
}

File: go/tree/recoverTree.go
package tree

func recoverTree(root *TreeNode) {
	var stack []*TreeNode
	curr := root
	var prev, x, y *TreeNode

	for curr != nil || len(stack) != 0 {
		for curr != nil {
			stack = append(stack, curr)
			curr = curr.Left
		}
		curr = stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		if prev != nil && prev.Val > curr.Val {
			y = curr
			if x == nil {
				x = prev
			} else {
				break
			}
		}
		prev = curr
		curr = curr.Right
	}
	swap(x, y)
}

func swap(x, y *TreeNode) {
	temp := x.Val
	x.Val = y.Val
	y.Val = temp
}
